### PR TITLE
feat: render LaTeX math in chat markdown (THU-606)

### PR DESCRIPTION
**Pull request by darkbanjo**
## What

Generalizes the LaTeX/math rendering prototyped on the OUP demo branch into a standard chat-markdown feature — no demo-specific coupling.

- Wires `remark-math` + `rehype-katex` into the shared `MemoizedMarkdown` renderer and ships the KaTeX stylesheet on the chat critical path.
- Adds `normalizeDisplayMath`: promotes a whole-line `$$…$$` equation to fenced display math (models routinely emit standalone equations on one line). Inline `$…$`, mid-sentence `$$…$$`, and already-fenced blocks are left untouched.
- Adds `katex`, `rehype-katex`, `remark-math` deps.

## Test

- New `memoized-markdown.test.tsx`: inline, display, mixed-with-GFM, single-line-amid-prose, and lone-dollar cases. 5/5 pass; typecheck clean.

Closes THU-606

---

> [!NOTE]
> **Low Risk**
> UI-only chat markdown rendering with defensive preprocessing; no auth, API, or data-path changes. Main risk is edge-case mis-rendering of `$` or math inside nested blockquotes/lists (documented limitations).
> 
> **Overview**
> Chat messages can now show **LaTeX math** (inline and display) instead of raw delimiters.
> 
> `MemoizedMarkdown` adds **`remark-math` + `rehype-katex`**, loads **KaTeX CSS**, and runs a **`normalizeDisplayMath` pipeline** before block splitting: `\[…\]` / `\(…\)` → `$` / `$$`, whole-line `$$…$$` → fenced display math (with list-indent preserved), and **currency `$` before digits** escaped so amounts like `$5` stay plain text. Fenced/indented code and inline `` `…` `` spans are skipped so literal math examples aren’t rewritten.
> 
> The **system prompt** now tells models to use `$…$` / `$$…$$` only. **`memoized-markdown.test.tsx`** covers GFM + math, delimiter variants, currency edge cases, and code-block safety.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5685709bbcad51fa692a0fe84e0366cffe5d767a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/components/chat/memoized-markdown.tsx [hunk 1]
```diff
@@ -6,20 +6,44 @@ import { marked } from 'marked'
...
+// single-line `.` keep the match to a whole line), and already-fenced blocks
+// don't match (their `$$` fences are alone on their lines).
+const displayMathLine = /^([ \t]*)\$\$[ \t]*(.+?)[ \t]*\$\$[ \t]*$/gm
+
+const normalizeDisplayMath = (markdown: string): string =>
+  markdown.replace(displayMathLine, (_match, indent: string, body: string) => `${indent}$$\n${body}\n$$`)
+
 const parseMarkdownIntoBlocks = (markdown: string): string[] => {
-  const tokens = marked.lexer(markdown)
+  const tokens = marked.lexer(normalizeDisplayMath(markdown))
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/components/chat/memoized-markdown.tsx]: Fixed in 0d90e700 — `parseMarkdownIntoBlocks` now lexes into top-level tokens first and skips `code` tokens (which covers both fenced *and* indented code) when rewriting math, then re-lexes so single-line `$$…$$` promotion still works. `normalizeDisplayMath` also leaves inline code spans verbatim, so `\(…\)` shown as inline code keeps its literal text too. Added tests covering a `$$…$$` line in a fenced block and `\(x\)` in an inline code span.

File: src/components/chat/memoized-markdown.tsx [hunk 2]
```diff
@@ -6,20 +6,81 @@ import { marked } from 'marked'
...
 
+// KaTeX styles for LaTeX math rendered by remark-math + rehype-katex. Imported
+// here so the stylesheet ships with the markdown renderer (chat critical path).
+import 'katex/dist/katex.min.css'
+
+// remark-math parses `$…$` / `$$…$$` into math nodes; rehype-katex renders them
+// to KaTeX HTML. Shared across every markdown block so inline and display math
+// render consistently. GFM stays first so its tokenizer runs before math.
+const remarkPlugins = [remarkGfm, remarkMath]
+const rehypePlugins = [rehypeKatex]
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/components/chat/memoized-markdown.tsx:23]: Fixed in 0ff813b6 — a currency `$` (start/whitespace/`(` followed by a digit) is now escaped to `\$` before remark-math parses, so "$5 and $10" renders as literal text instead of a math span. `$$` fences aren't matched, and real inline math like `$x^2$` in the same sentence still renders. Added tests for two amounts in one sentence and for currency mixed with real math.

File: src/components/chat/memoized-markdown.tsx [hunk 3]
```diff
@@ -6,20 +6,81 @@ import { marked } from 'marked'
...
+// don't match (their `$$` fences are alone on their lines).
+const displayMathLine = /^([ \t]*)\$\$[ \t]*(.+?)[ \t]*\$\$[ \t]*$/gm
+
+// Rewrite the math delimiters in a span of prose. Never called on code — see
+// `normalizeDisplayMath` (skips inline code spans) and `parseMarkdownIntoBlocks`
+// (skips fenced/indented code blocks).
+const rewriteMath = (text: string): string =>
+  text
+    .replace(displayMathDelimiters, (_match, body: string) => `$$\n${body.trim()}\n$$`)
+    .replace(inlineMathDelimiters, (_match, body: string) => `$${body.trim()}$`)
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/components/chat/memoized-markdown.tsx]: Fixed in 0ff813b6 — `inlineMathDelimiters` now uses `[\s\S]+?` (matching the display pattern), so a `\(…\)` split across lines is converted, with internal whitespace collapsed to a single space so it renders as clean inline math. Added a multi-line `\(…\)` test.

File: src/components/chat/memoized-markdown.tsx [hunk 4]
```diff
@@ -6,20 +6,94 @@ import { marked } from 'marked'
...
-  const tokens = marked.lexer(markdown)
-  return tokens.map((token) => token.raw)
+  // Rewrite math per top-level token, skipping `code` tokens (fenced *and*
+  // indented code) so a message that shows `$$…$$` as source keeps its literal
+  // text. Re-lex the rewritten string so a promoted single-line `$$…$$` still
+  // splits into its own display-math block.
+  const normalized = marked
+    .lexer(markdown)
+    .map((token) => (token.type === 'code' ? token.raw : normalizeDisplayMath(token.raw)))
+    .join('')
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/components/chat/memoized-markdown.tsx:157]: Acknowledged as a known, deliberately-scoped limitation. Top-level code blocks (fenced + indented) are skipped, and inline code spans are protected at any nesting depth; a fenced/indented code block nested inside a blockquote or list can still have its LaTeX/currency text rewritten, because marked strips the container prefixes from the nested token's `raw` so it can't be matched back to the source. A fully code-aware pass would need AST-level handling. Given how rare this nested case is in chat, we've documented it in `parseMarkdownIntoBlocks` (7a9d2a19) and are accepting it rather than rewriting the renderer.

File: src/components/chat/memoized-markdown.tsx [hunk 5]
```diff
@@ -6,20 +6,104 @@ import { marked } from 'marked'
...
+// `normalizeDisplayMath` (skips inline code spans) and `parseMarkdownIntoBlocks`
+// (skips fenced/indented code blocks).
+const rewriteMath = (text: string): string =>
+  text
+    .replace(displayMathDelimiters, (_match, body: string) => `$$\n${body.trim()}\n$$`)
+    // Inline math is single-line by nature, so collapse any internal whitespace
+    // (including the newline a multi-line `\(…\)` carried) to a single space.
+    .replace(inlineMathDelimiters, (_match, body: string) => `$${body.trim().replace(/\s+/g, ' ')}$`)
+    .replace(displayMathLine, (_match, indent: string, body: string) => `${indent}$$\n${body}\n$$`)
+    .replace(currencyDollar, (_match, lead: string) => `${lead}\\$`)
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/components/chat/memoized-markdown.tsx]: Acknowledged as a deliberately-scoped edge. A `\(…\)` nested inside display math (`\[…\]` / `$$…$$`) is malformed LaTeX — math modes don't nest — so the only inputs affected here are already invalid; well-formed equations are never touched. We've documented the scope in `rewriteMath` (6cf65e3f) rather than add special-casing to the renderer for input that can't render correctly either way.

File: package.json [hunk 6]
```diff
@@ -114,6 +114,7 @@
     "drizzle-orm": "^0.45.2",
     "framer-motion": "^12.23.12",
     "input-otp": "^1.4.2",
+    "katex": "^0.17.0",
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply package.json]: Fixed in 355d9a32 — pinned the direct `katex` dep to `^0.16.0` so the imported stylesheet (`katex/dist/katex.min.css`) matches the `0.16.47` renderer that rehype-katex / micromark-extension-math use. The lockfile now resolves a single `katex@0.16.47`, so the CSS and the rendered markup are on the same version. Good catch on the `@types/katex` hint.

File: src/components/chat/memoized-markdown.tsx [hunk 7]
```diff
@@ -6,20 +6,128 @@ import { marked } from 'marked'
...
+const rewriteMath = (text: string): string =>
+  // Currency escaping runs last, on the fully-normalized text, so it sees every
+  // `$$…$$` / `$…$` span — including those just converted from `\[…\]` / `\(…\)`.
+  escapeCurrencyDollars(
+    text
+      .replace(displayMathDelimiters, (_match, body: string) => `$$\n${body.trim()}\n$$`)
+      // Inline math is single-line by nature, so collapse internal whitespace
+      // (including the newline a multi-line `\(…\)` carried) to a single space.
+      .replace(inlineMathDelimiters, (_match, body: string) => `$${body.trim().replace(/\s+/g, ' ')}$`)
+      .replace(displayMathLine, (_match, indent: string, body: string) => `${indent}$$\n${body}\n$$`),
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/components/chat/memoized-markdown.tsx]: Fixed in 355d9a32 — added an `indentedFence` helper so promoting a one-line `$$…$$` or `\[…\]` equation preserves the line's leading indentation on the body and the closing fence, not just the opening one. An indented (list/blockquote-continuation) equation now stays inside its item instead of dedenting to column 0 and ending the list early. Added tests for `$$` and `\[…\]` inside a list item (display math renders, both list items intact).

File: src/components/chat/memoized-markdown.tsx [hunk 8]
```diff
@@ -6,20 +6,145 @@ import { marked } from 'marked'
...
+const inlineMathDelimiters = /\\\(([\s\S]+?)\\\)/g
+
+// remark-math only renders `$$…$$` as centered *display* math when the fences
+// sit on their own lines; a single-line `$$…$$` falls back to inline. Models
+// routinely emit standalone equations on a single line, so rewrite any line
+// that is wholly a `$$…$$` equation into the fenced form.
+// Inline `$…$` and mid-sentence `$$…$$` are left untouched (the `$` anchors and
+// single-line `.` keep the match to a whole line), and already-fenced blocks
+// don't match (their `$$` fences are alone on their lines).
+const displayMathLine = /^([ \t]*)\$\$[ \t]*(.+?)[ \t]*\$\$[ \t]*$/gm
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/components/chat/memoized-markdown.tsx:48]: Acknowledged as a documented limitation (a0ca8f0f). Display-math promotion preserves whitespace (list-continuation) indentation but not a blockquote's `>` line prefix, so a `\[…\]` / single-line `$$…$$` *inside a blockquote* isn't promoted — same container-context class as the nested-code note in `parseMarkdownIntoBlocks`. It's rare in chat (display math in a quote/callout), and a fully container-safe fix means moving delimiter handling to a remark AST plugin, which we'd take on separately if container math ever shows up. Inline `\(…\)` still works in blockquotes since it isn't line-anchored.

File: src/components/chat/memoized-markdown.tsx [hunk 9]
```diff
@@ -6,20 +6,153 @@ import { marked } from 'marked'
...
+      .replace(inlineMathDelimiters, (_match, body: string) => `$${body.trim().replace(/\s+/g, ' ')}$`)
+      .replace(displayMathLine, (_match, indent: string, body: string) => indentedFence(indent, body)),
+  )
+
+// An inline code span (`` `…` ``, `` ``…`` ``, …). Left verbatim so a message
+// that shows `$$…$$` / `\(…\)` *as inline code* keeps its literal text.
+const inlineCodeSpan = /(`+)[^`]*?\1/g
+
+// Rewrite math everywhere except inside inline code spans.
+const normalizeDisplayMath = (markdown: string): string => {
```
Comment (ital0) [inline src/components/chat/memoized-markdown.tsx:129]: Nice job getting all these edge cases covered! One thing worth a thought: this layer rewrites the raw markdown with regexes before lexing, and you've already spotted two cases it can't quite reach (display math in a blockquote, and code nested inside a list or blockquote). Those would mostly go away if the delimiter conversion ran as a small remark transform over text nodes, since that skips code and respects containers for free. Did you weigh that route? Flagging it mainly because this feels like the spot we'll keep patching as new shapes show up.

File: src/components/chat/memoized-markdown.tsx [hunk 10]
```diff
@@ -6,20 +6,153 @@ import { marked } from 'marked'
...
 import ReactMarkdown from 'react-markdown'
+import rehypeKatex from 'rehype-katex'
 import remarkGfm from 'remark-gfm'
+import remarkMath from 'remark-math'
 
 import { markdownComponents } from './markdown-utils'
 
+// KaTeX styles for LaTeX math rendered by remark-math + rehype-katex. Imported
+// here so the stylesheet ships with the markdown renderer (chat critical path).
+import 'katex/dist/katex.min.css'
```
Comment (ital0) [inline src/components/chat/memoized-markdown.tsx:17]: can we lazy load this?

File: src/components/chat/memoized-markdown.tsx [hunk 11]
```diff
@@ -6,20 +6,153 @@ import { marked } from 'marked'
...
+  // Known limitation: only *top-level* code blocks are skipped here. A fenced or
+  // indented code block nested inside a blockquote/list lives in the parent
+  // token's `raw` (marked strips its container prefixes, so it can't be matched
+  // back to the source), and `rewriteMath` still runs on it — a message that
+  // shows `$$…$$`/`\(…\)` as code *inside a blockquote or list* may have that
+  // text rewritten. Inline code spans (`` `…` ``) are protected at any depth via
+  // `normalizeDisplayMath`. This nested-block case is rare in chat; a fully
+  // code-aware pass would need AST-level handling and is deliberately out of
+  // scope here.
+  const normalized = marked
```
Comment (ital0) [inline src/components/chat/memoized-markdown.tsx:154]: Small one: this now runs marked.lexer twice per render (once to normalize, then again to split into blocks). It's memoized on content, but while streaming, content changes every token, so we pay for both lexes on each delta. Likely fine for normal message sizes, just wanted to make sure it's a deliberate trade rather than a surprise.

File: src/ai/prompt.ts [hunk 12]
```diff
@@ -124,5 +124,6 @@ Wrong: "The metro area has 37 million residents.\n[1]" (citation on new line)
 Wrong: "Tokyo has 14 million residents. [1] The metro area has 37 million. [1]" (repeated [1])
 Wrong: "Tokyo has 14 million residents." (missing [N])
 Wrong: "| Tokyo | 14 million | [1] |" (citation in separate column)
+Format math as LaTeX with dollar delimiters: $…$ inline, $$…$$ for standalone equations. Never use \\(…\\) or \\[…\\].
```
Comment (ital0) [inline src/ai/prompt.ts:131]: Quick intent question: this tells the model to never use \(…\) or \[…\], but the renderer goes to real lengths to support exactly those. Totally reasonable as a safety net since models drift, but the two read like they disagree. Maybe a short note here (something like "we ask for $ only but still normalize \( and \[ defensively") so nobody later removes one side thinking it's redundant.

Does it makes sense?

**Review by ital0 — CHANGES_REQUESTED** [review]

File: src/components/chat/memoized-markdown.tsx (hunk 9 above)
Comment (darkbanjo) [reply src/components/chat/memoized-markdown.tsx:129]: Good call, and we did weigh it. The catch: the currency-escape and the single-line `$$…$$`→display promotion must run *before* remark-math tokenizes `$`, which a post-parse transform can't do — so only the `\(…\)`/`\[…\]` conversion moves to a transform; the rest stays pre-parse. So it's a partial rewrite, not a free swap. Agreed it's where we keep patching, so I've tracked the refactor as THU-636 (it'd also close the blockquote + nested-code limitations and the double-lex note) rather than land it in this PR.

File: src/components/chat/memoized-markdown.tsx (hunk 10 above)
Comment (darkbanjo) [reply src/components/chat/memoized-markdown.tsx:17]: Agreed — it's ~78 KB gzipped on the chat critical path. Tracked as THU-635. Deferring out of this PR because lazy-loading forces the renderer to go async (a brief raw-`$` flash until the chunk loads + rewriting the ~10 synchronous markdown tests to async, which hits the repo's waitFor/fake-timer quirk) — worth its own focused PR rather than rushing into final review.

File: src/components/chat/memoized-markdown.tsx (hunk 11 above)
Comment (darkbanjo) [reply src/components/chat/memoized-markdown.tsx:154]: Deliberate, yep. The second lex is needed because normalization can change block structure (a one-line `$$…$$` becomes a fenced multi-line block), so the blocks must be re-derived after rewrite. Memoized on content, so it only re-runs as content grows while streaming; fine for normal message sizes. The remark-transform refactor (THU-636) would remove the pre-pass entirely and with it the double lex — flagged there.

File: src/ai/prompt.ts (hunk 12 above)
Comment (darkbanjo) [reply src/ai/prompt.ts:131]: Done in a5cad9d3 — added cross-referencing comments in both directions (prompt.ts ↔ memoized-markdown.tsx) noting the `$`-only instruction and the defensive `\(…\)`/`\[…\]` normalization are complementary, not redundant, so neither side gets removed as dead.

File: src/components/chat/memoized-markdown.tsx [hunk 13]
```diff
@@ -6,20 +6,156 @@ import { marked } from 'marked'
...
+  // Currency escaping runs last, on the fully-normalized text, so it sees every
+  // `$$…$$` / `$…$` span — including those just converted from `\[…\]` / `\(…\)`.
+  escapeCurrencyDollars(
+    text
+      .replace(displayMathDelimiters, (_match, indent: string | undefined, body: string) =>
+        indentedFence(indent ?? '', body),
+      )
+      // Inline math is single-line by nature, so collapse internal whitespace
+      // (including the newline a multi-line `\(…\)` carried) to a single space.
+      .replace(inlineMathDelimiters, (_match, body: string) => `$${body.trim().replace(/\s+/g, ' ')}$`)
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/components/chat/memoized-markdown.tsx]: Fixed in 5685709b — both conversions now guard against an empty (whitespace-only) body: `\( \)` and `\[ \]` are left as literal text instead of being rewritten to a bare `$$`, so a stray empty inline can't open a display fence and swallow following prose. (`\(\)` with nothing between already didn't match — the pattern needs ≥1 char — this covers the whitespace case.) Added a test that an empty `\( \)` in prose renders no math and doesn't eat the text after it.
